[ARM][SDAG] Half promote llvm.lrint nodes.

davemgreen opened the pull request:
As shown in #137101, fp16 lrint are not handled correctly on Arm. This adds soft-half promotion for them, reusing the function that promotes a value with operands (and can handle strict fp once that is added).

RKSimon reviewed (changes requested):
please can you regenerate the test files first and then rebase? its difficult to see whats the codegen change and whats the refactor

davemgreen committed:
[ARM][SDAG] Half promote llvm.lrint nodes.

As shown in #137101, fp16 lrint are not handled correctly on Arm. This adds
soft-half promotion for them, reusing the function that promotes a value with
operands (and can handle strict fp once that is added).

davemgreen committed:
Make the fp16 run line use +fp-armv8

efriedma-quic commented on llvm/test/CodeGen/ARM/lrint-conv.ll:
```diff
@@ -3,12 +3,35 @@
 ; RUN: llc < %s -mtriple=armv7-none-eabihf -mattr=+vfp2 -float-abi=hard | FileCheck %s --check-prefixes=CHECK,CHECK-NOFP16
 ; RUN: llc < %s -mtriple=armv7-none-eabihf -mattr=+vfp2,+fullfp16 -float-abi=hard | FileCheck %s --check-prefixes=CHECK,CHECK-FP16
```
I'm specifically wondering about the case where we have fparmv8 but not fullfp16; if I'm understanding correctly, we should be able to __aeabi_h2f to float, and then vrintx.f32.  Want to see that's happening, as opposed to somehow calling lrintf.

davemgreen replied:
Ah I see what you mean, I've added it in. It does the opposite as vfp4 includes the original fp16 convert instructions, but we don't use vrintx.f32 for the f32. It looks like the f16 is type legalized before the lrint is expanded.

davemgreen committed:
Add a fparmv8 run line

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -3740,7 +3740,10 @@ bool DAGTypeLegalizer::SoftPromoteHalfOperand(SDNode *N, unsigned OpNo) {
   case ISD::FP_TO_SINT:
-  case ISD::FP_TO_UINT: Res = SoftPromoteHalfOp_FP_TO_XINT(N); break;
+  case ISD::FP_TO_UINT:
+  case ISD::LRINT:
```
Also handle llvm.llrint while you're here?

davemgreen committed:
Also add llrint, whilst I am here

efriedma-quic reviewed (approved):
LGTM

(Like I mentioned, we probably want to mark more variants of lrint "Expand", but we can leave that for a followup.)

nikic commented:
Related PR: https://github.com/llvm/llvm-project/pull/152684

davemgreen commented:
> Like I mentioned, we probably want to mark more variants of lrint "Expand", but we can leave that for a followup.

Yep, I agree. I was trying to keep this focussed on fixing lrint. (And some of the llrint cases were not trivial, due to the way it legalizes llrint). I have another (simple) patch for lround too, that is essentially the same as this. I haven't looked at strict nodes, Arm doesn't support them quite yet.
